src/transformations/add-storage-gaps.ts

```diff
@@ -1,66 +1,115 @@
 import { SourceUnit, ContractDefinition } from 'solidity-ast';
-import { findAll } from 'solidity-ast/utils';
+import { findAll, isNodeType } from 'solidity-ast/utils';
 
 import { formatLines } from './utils/format-lines';
+import { hasOverride } from '../utils/upgrades-overrides';
 import { getNodeBounds } from '../solc/ast-utils';
 import { StorageLayout } from '../solc/input-output';
 import { Transformation } from './type';
 import { TransformerTools } from '../transform';
+import { extractNatspec } from '../utils/extractNatspec';
+import { decodeTypeIdentifier } from '../utils/type-id';
 
-// 100 slots of 32 contractSize each
-const TARGET_SIZE = 32 * 50;
+// By default, make the contract a total of 50 slots (storage + gap)
+const DEFAULT_SLOT_COUNT = 50;
 
 export function* addStorageGaps(
   sourceUnit: SourceUnit,
   { getLayout }: TransformerTools,
 ): Generator<Transformation> {
   for (const contract of findAll('ContractDefinition', sourceUnit)) {
     if (contract.contractKind === 'contract') {
-      const gapSize = getGapSize(contract, getLayout(contract));
+      let targetSlots = DEFAULT_SLOT_COUNT;
+      for (const entry of extractNatspec(contract)) {
+        if (entry.title === 'custom' && entry.tag === 'storage-size') {
+          targetSlots = parseInt(entry.args);
+        }
+      }
 
-      const contractBounds = getNodeBounds(contract);
-      const start = contractBounds.start + contractBounds.length - 1;
+      const gapSize = targetSlots - getContractSlotCount(contract, getLayout(contract));
 
-      const text = formatLines(0, [
-        ``,
-        [
-          `/**`,
-          ` * @dev This empty reserved space is put in place to allow future versions to add new`,
-          ` * variables without shifting down storage in the inheritance chain.`,
-          ` * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps`,
-          ` */`,
-          `uint256[${gapSize}] private __gap;`,
-        ],
-      ]);
+      if (gapSize > 0) {
+        const contractBounds = getNodeBounds(contract);
+        const start = contractBounds.start + contractBounds.length - 1;
 
-      yield {
-        kind: 'add-storage-gaps',
-        start,
-        length: 0,
-        text,
-      };
+        const text = formatLines(0, [
+          ``,
+          [
+            `/**`,
+            ` * @dev This empty reserved space is put in place to allow future versions to add new`,
+            ` * variables without shifting down storage in the inheritance chain.`,
+            ` * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps`,
+            ` */`,
+            `uint256[${gapSize}] private __gap;`,
+          ],
+        ]);
+
+        yield {
+          kind: 'add-storage-gaps',
+          start,
+          length: 0,
+          text,
+        };
+      } else if (gapSize < 0) {
```

Comment:
```suggestion
      } else if (gapSize <= 0) {
```

Gap size should be at least 1, because we also need the gap to reset alignment: the next contract in inheritance should start its layout in a new slot. This is clearly seen in the test contract `SizeOverrideExact`: if the last variable is a bool, the next one may be packed with it.